### PR TITLE
Support building fully-static cc libraries

### PR DESCRIPTION
**Pull request by ianull**

### CODE REVIEW

File: cc/build_defs.bzl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+"""Provides a rule that outputs a monolithic static library."""
+
+load(
+    "@bazel_tools//tools/cpp:toolchain_utils.bzl",
+    "find_cpp_toolchain",
+    "use_cpp_toolchain",
+)
+
+# Bazel rule for collecting the header files that a target depends on.
```
Comment (conradgrobler) [inline cc/build_defs.bzl:25]: nit: I don't think that this comment accurately describes everything that is done here.

**Review by conradgrobler — APPROVED** [review]
I am not that familiar with build rules of the details of AR scripts, but the implementation looks good as far as I can tell. Thanks.

File: cc/build_defs.bzl (hunk 1 above)
Comment (ipetr0v) [inline cc/build_defs.bzl:25]: Could you please add a link where this code comes from?

File: cc/build_defs.bzl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+def _cc_static_library_impl(ctx):
```
Comment (ipetr0v) [inline cc/build_defs.bzl:26]: How does this rule handle [copts](https://bazel.build/reference/be/c-cpp#cc_library.copts) and [defines](https://bazel.build/reference/be/c-cpp#cc_library.defines) ?

File: cc/build_defs.bzl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            for flag in inp.user_link_flags
+        })
+    link_flags = unique_flags.keys()
+
+    # Collect static libraries
+    libs = []
+    for inp in input_depset.to_list():
+        for lib in inp.libraries:
+            if lib.pic_static_library:
+                libs.append(lib.pic_static_library)
```
Comment (ipetr0v) [inline cc/build_defs.bzl:52]: How does it handle duplicated dependencies?

File: cc/build_defs.bzl [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                libs.append(lib.static_library)
+
+    script_file = ctx.actions.declare_file("lib{}.mri".format(ctx.attr.name))
+    commands = ["create {}".format(output_lib.path)]
+    for lib in libs:
+        commands.append("addlib {}".format(lib.path))
+    commands.append("save")
+    commands.append("end")
+    ctx.actions.write(
+        output = script_file,
```
Comment (ipetr0v) [inline cc/build_defs.bzl:63]: In our CI we always check that we don't generate new files in the repo.
Does this rule create a new file that is not tracked by the repository?

Comment (tiziano88) [reply cc/build_defs.bzl:63]: These are bazel rules, I think their only output is within the bazel folders, they cannot possibly write to random parts of the host file system (I hope).

File: cc/build_defs.bzl (hunk 3 above)
Comment (tiziano88) [reply cc/build_defs.bzl:52]: It should not matter if the libs are specified multiple times I think.

**Review by tiziano88 — APPROVED** [review]

**Review by ipetr0v — APPROVED** [review]

**Review by andrisaar — APPROVED** [review]

File: cc/build_defs.bzl (hunk 1 above)
Comment (ianull) [reply cc/build_defs.bzl:25]: lol, should be **object files** instead of **header files**.

Comment (ianull) [reply cc/build_defs.bzl:25]: np

File: cc/build_defs.bzl (hunk 2 above)
Comment (ianull) [reply cc/build_defs.bzl:26]: It doesn't need to handle copts and defines as it's working as a wrapper :)

Use current codebase as example:
```
cc_library(
    name = "tflite-micro",
    srcs = ["tflite_micro.cc"],
    hdrs = ["tflite_micro.h"],
    linkstatic = 1,
    deps = [
        "@com_github_tensorflow_tflite_micro//tensorflow/lite/micro:micro_framework",
    ],
)

cc_static_library(
    name = "tflite-micro-fully-static",
    deps = [":tflite-micro"],
)
```
The `tflite-micro-fully-static` target searches all required libraries for `tflite-micro`, then uses `ar` to archive all of them into one archive a.k.a. fully static library.
In this case, all *.a generated from tflite-micro as well as @com_github_tensorflow_tflite_micro//tensorflow/lite/micro:micro_framework are `ar` into a fully static `libtflite-micro-fully-static.a`.

File: cc/build_defs.bzl (hunk 3 above)
Comment (ianull) [reply cc/build_defs.bzl:52]: The `unique` magic is done by `depset`: `input_depset = depset(transitive = lib_sets)`

File: cc/build_defs.bzl (hunk 4 above)
Comment (ianull) [reply cc/build_defs.bzl:63]: Like what Tiziano said: the files (intermediates and final static .a lib) the rule generates are put in bazel-bin/**/* and not touching any of host files.

**Comment by ianull** [comment]
How do I get pass the error that is blocking the merge:
`= note: /usr/bin/ld: final link failed: No space left on device`

@tiziano88 could you help clean up device space?

**Comment by tiziano88** [comment]
It looks like that's from a previous run, so hopefully it will succeed now, I think @conradgrobler cleared the cache earlier today, so it should leave more space for the build. But let me know if still fails

**Pull request merged**